blocks/serialization.py

```diff
@@ -33,142 +169,93 @@
-    number of external objects. Note that the zip files are compatible with
-    NumPy's :func:`numpy.load` function.
-
-    >>> import numpy
-    >>> from blocks.bricks import MLP, Identity
-    >>> from blocks.initialization import Constant
-    >>> mlp = MLP([Identity()], [10, 10], weights_init=Constant(0.),
-    ...           biases_init=Constant(0.))
-    >>> mlp.initialize()
-    >>> with open('model.zip', 'wb') as f:
-    ...     dump(mlp, f)
-    >>> 'mlp-linear_0.W' in numpy.load('model.zip').keys()
-    True
-    >>> 'mlp-linear_0.b' in numpy.load('model.zip').keys()
-    True
-    >>> numpy.load('model.zip')['mlp-linear_0.W'].shape
-    (10, 10)
-    >>> with open('model.zip', 'rb') as f:
-    ...     mlp2 = load(f)
-    >>> mlp2  # doctest: +ELLIPSIS
-    <blocks.bricks.MLP object at ...: name=mlp>
+        Use cPickle instead of pickle. Setting it to true will disable the
+        warning message if you try to pickle objects from the main module!
+        Be sure that you don't have the warning before turning this flag
+        on. Default: False.
+    \*\*kwargs
+        Keyword arguments to be passed to `pickle.Pickler`.
 
     """
-    with closing(zipfile.ZipFile(file_handler, 'w', zipfile.ZIP_DEFLATED,
-                                 allowZip64=True)) as zip_file:
-        def func(f):
-            if use_cpickle:
-                p = cPickle.Pickler(f, protocol=protocol)
-            else:
-                p = PicklerWithWarning(f, protocol=protocol)
-            p.persistent_id = persistent_id(zip_file)
-            p.dump(obj)
-        pkl_utils.zipadd(func, zip_file, 'pkl')
-
-
-# A thin wrapper around Theano load.
-load = pkl_utils.load
-
-
-def secure_dump(object_, path, dump_function=dump, **kwargs):
+    if not pickle_separately:
+        pickle_separately = {}
+    if '_pkl' in pickle_separately or '_parameters' in pickle_separately:
+        raise ValueError("_pkl and _parameters are reserved names and can't" \
+                         " be used as keys in pickle_separately.")
+
+    if use_cpickle:
+        pickler = cPickle.Pickler
+    else:
+        pickler = _PicklerWithWarning
+
+    with closing(tarfile.TarFile(fileobj=file_, mode='w')) as tar_file:
+        external_objects = {}
+        def _save_parameters(f):
+            renamer = _Renamer()
+            named_parameters = {renamer(p): p for p in parameters}
+            numpy.savez(f, **{name: p.get_value()
+                              for name, p in named_parameters.items()})
+
+            for name, p in named_parameters.items():
+                array_ = p.container.storage[0]
+                external_objects[id(array_)] = _mangle_parameter_name(
+                    type(array_), name)
+        if parameters:
+            _taradd(_save_parameters, tar_file, '_parameters')
+        if pickle_separately:
+            for name, component in six.iteritems(pickle_separately):
+                def _pickle_separately(f):
```

Comment:
As we discussed, the fact that we do not add ids of the objects to  `external_objects` here means they will be pickled together with the whole object. We should decide if we are fine with that, and if yes, change the docstring of the module.